include/pybind11/pybind11.h

```diff
@@ -827,18 +827,19 @@ class module : public object {
     PYBIND11_OBJECT_DEFAULT(module, object, PyModule_Check)
 
     /// Create a new top-level Python module with the given name and docstring
-    explicit module(const char *name, const char *doc = nullptr) {
-        if (!options::show_user_defined_docstrings()) doc = nullptr;
 #if PY_MAJOR_VERSION >= 3
-        PyModuleDef *def = new PyModuleDef();
+    explicit module(const char *name, const char *doc = nullptr, PyModuleDef *def = nullptr) {
+        if (!def) def = new PyModuleDef();
         std::memset(def, 0, sizeof(PyModuleDef));
```

Comment:
Is this `memset` required? `new T()` is doing *value* initialization, so any C struct would already be set to zero.

Except there's the `static` one, which is *default* initialized, leaving everything in uninitialized state. Can we do `static PyModuleDef def();`? Trying...

As expected, "the most vexing parse" is making things difficult. `static auto def = PyModuleDef();` works.

"But Boris... that's *copy* initialization! Copying is in the name... Copy *bad*!"

Short answer - check the assembly.

Long answer - Copy elision isn't a normal optimization. It has to be performed very early if it is going to be performed at all. That's why copy elision is "visible" at `-O0`.

&nbsp;

Anyway, totally non-blocking.

Reply:
Actually, `static auto def = PyModuleDef();` produces this directive:

```asm
function_name():
	; snip
	.type function_name()::def,@object
	.local function_name()::def
	.comm function_name()::def,104,32 ; common subsection of .data, 104 uninitialized bytes, 32byte alignment.
	                                  ; It's not in .bss, which is initialized to 0 for free
```

Therefore, we need the `memset` somewhere. The `new PyModuleDef()` does initialize the struct to zero.